compiler/passes/denormalize.cpp

```diff
@@ -0,0 +1,457 @@
+void denormalizeOrDeferCandidates(UseDefCastMap& candidates,
+    Vec<Symbol*>& deferredSyms);
+
+/*
+ * This function tries to remove temporary variables in function `fn`
+ *
+ * A local variable is removed if:
+ *
+ * - It is def'd and use'd once
+ *
+ * - Its def is a PRIM_MOVE or PRIM_ASSIGN, with no possible communication
+ * - RHS and LHS are of same type and non-extern
+ *
+ * - Its use a suitable primitive
+ * - Its use is not repeated(condition/increment statement of a loop)
+ *
+ * Denormalization uses helpers in util/exprAnalysis.cpp to decide if it's safe
+ * to move function calls and primitives.
+ */
+void denormalize(void) {
+
+  UseDefCastMap candidates;
+  Vec<Symbol*> deferredSyms;
+
+  if(fDenormalize) {
+    forv_Vec(FnSymbol, fn, gFnSymbols) {
+      bool isFirstRound = true;
+      do {
+        candidates.clear();
+        deferredSyms.clear();
+
+        // if we are analyzing locals in a function for the first time, we look
+        // at all the symbols in the function. otherwise we only look at the
+        // ones deferred in the previous passes on the same function
+        if(isFirstRound) {
+          findCandidatesInFunc(fn, candidates);
+        }
+        else {
+          findCandidatesInFunc(fn, deferredSyms, candidates);
+        }
+        denormalizeOrDeferCandidates(candidates, deferredSyms);
+
+        isFirstRound = false;
+      } while(deferredSyms.count() > 0);
+    }
+  }
+}
+
+void denormalizeOrDeferCandidates(UseDefCastMap& candidates,
+    Vec<Symbol*>& deferredSyms) {
+
+  for(UseDefCastMap::iterator it = candidates.begin() ;
+      it != candidates.end() ; ++it) {
+    // unpack the bundle
+    std::pair<Expr*, Type*> defCastPair = it->second;
+    SymExpr* use = it->first;
+    Expr* def = defCastPair.first;
+    Type* castTo = defCastPair.second;
+
+    // if parent of def is gone, it means it has been denormalized in the
+    // earlier passes
+    if(def->parentExpr == NULL) {
+      deferredSyms.add(use->var);
+      continue;
+    }
+    denormalize(def, use, castTo);
+  }
+}
+
+void findCandidatesInFunc(FnSymbol* fn, Vec<Symbol*> symVec,
+    UseDefCastMap& udcMap) {
+
+  Map<Symbol*,Vec<SymExpr*>*> defMap;
+  Map<Symbol*,Vec<SymExpr*>*> useMap;
+
+  buildDefUseMaps(fn, defMap, useMap);
+
+  bool cachedGlobalManip = isRegisteredGlobalManip(fn);
+  // Note the different symbol set being iterated
+  forv_Vec(Symbol, sym, symVec) {
+
+    SymExpr *use = NULL;
+    Expr *usePar = NULL;
+    Expr *def = NULL;
+    Expr *defPar;
+    Type* castTo = NULL;
+
+    //if we don't already have it cached,
+    //check for global symbols in function body
+    if(!cachedGlobalManip) {
+      //I am not sure if there should be !isConstant and !isParam as well
+      //I think there can be a const but stateful global that is manipulated by
+      //the function so, play safe
+      if(sym && !sym->isImmediate() && isGlobal(sym)){
+        registerGlobalManip(fn, true);
+        cachedGlobalManip = true;
```

Comment:
Isn't this code already in exprAnalysis.cpp ? Did you just forget to remove it from here?